src/db.rs

```diff
@@ -260,27 +262,60 @@ impl DbInner {
         }
     }
 
-    async fn replay_wal(&self) -> Result<(), SlateDBError> {
+    async fn replay_wal(self: &Arc<Self>) -> Result<(), SlateDBError> {
         let wal_id_last_compacted = self.state.read().state().core.last_compacted_wal_sst_id;
         let wal_sst_list = self
             .table_store
             .get_wal_sst_list(wal_id_last_compacted)
             .await?;
         let mut last_sst_id = wal_id_last_compacted;
+        let mut wal_tasks = Vec::new();
+        // we use a BTreeMap here to preserve order of sst_id
+        let mut wal_replay_buffers = BTreeMap::new();
+        let mut wal_replay_errors = Vec::new();
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(
+            self.options.maximum_parallel_tasks_limit,
+        ));
+
+        // constructs a list of tokio tasks to asynchronously fetch wal sst
         for sst_id in wal_sst_list {
-            last_sst_id = sst_id;
-            let sst = self.table_store.open_sst(&SsTableId::Wal(sst_id)).await?;
-            let sst_id = match &sst.id {
-                SsTableId::Wal(id) => *id,
-                SsTableId::Compacted(_) => return Err(SlateDBError::InvalidDBState),
-            };
-            let mut iter = SstIterator::new(&sst, self.table_store.clone(), 1, 1, true).await?;
-            // iterate over the WAL SSTs in reverse order to ensure we recover in write-order
-            // buffer the WAL entries to bulk replay them into the memtable.
-            let mut wal_replay_buf = Vec::new();
-            while let Some(kv) = iter.next_entry().await? {
-                wal_replay_buf.push(kv);
+            let this = self.clone();
+            let semaphore = semaphore.clone();
+            wal_tasks.push(tokio::spawn(async move {
+                match semaphore.acquire().await {
+                    Ok(_) => this.fetch_wal(sst_id).await,
```

Comment:
(question coming from total newbie) how big do we expect each WAL to be in memory? the previous approach would allow us to free memory from each SST after inserting it into the memtables while this one holds all of them in memory before freeing any of them (requiring up to 2x the total memory). 

Not sure if this is a problem or not, but if it is we should consider pipelining the execution so that we fetch N WAL SSTs, then insert them into the memtable before fetching the following N SSTs to avoid storing all the SSTs in memory at once (in addition to having them in the memtable).

Reply:
IIRC, right now, we only flush WALs on `flush_interval`; they're not size-bounded. I opened #30 early on to consider adding size support, but opted against it. Implicitly, the garbage collector will push back on L0 writes, and if the `max_unflushed_memtable` threshold is reached, the WAL writes will eventually be stopped. But if someone writes a lot of data in a single `flush_interval` window, the WAL SSTs could be quite large.

> we should consider pipelining the execution so that we fetch N WAL SSTs, then insert them into the memtable before fetching the following N SSTs

+1 Let's just go with 4. This doesn't have to be perfect right now.

Reply:
IIRC, right now, we only flush WALs on `flush_interval`; they're not size-bounded. I opened #30 early on to consider adding size support, but opted against it. Implicitly, the garbage collector will push back on L0 writes, and if the `max_unflushed_memtable` threshold is reached, the WAL writes will eventually be stopped. But if someone writes a lot of data in a single `flush_interval` window, the WAL SSTs could be quite large.

> we should consider pipelining the execution so that we fetch N WAL SSTs, then insert them into the memtable before fetching the following N SSTs

+1 Let's just go with 4. This doesn't have to be perfect right now.

Reply:
Sorry, yes compactor not GC. 😅

Reply:
so as far as i could understand my previous code achieved wal file level parallelization as was expected from the issue but the main concern was memory consumption due to all WALs being in one memory all at once, so I have made the following changes:
- Create `tokio::tasks` in chunks of 4 and loads wal into memory, build memtable using the 4 and then move onto the chunk of next 4
- because of the above change semaphore is no longer needed so i removed it
- replaced `SstIterator::new` with `SstIterator::new_spawn` using `max_fetch_task` as 1 and 256 blocks to leverage block level parallelism along with the file level parallelism introduced by this PR

Let me know if I understood the comments correctly cc @criccomini